### ISSUE TITLE
Add support for dynamically loading grammar libraries at runtime

### ISSUE THREAD

**Issue by jirutka** [Wilfred/difftastic#123]
Can you please add support for dynamically loading (any) grammar libraries at runtime? The current approach of vendoring every existing grammar and linking them in build-time is not scalable and not compatible with linux distributions.

This issue has been already solved e.g. in [diffsitter](https://github.com/afnanenayet/diffsitter) (https://github.com/afnanenayet/diffsitter/pull/177) or [helix](https://github.com/helix-editor/helix) (https://github.com/helix-editor/helix/pull/432).

You can either load a grammar library for the specified language from the system library path (this is typically `/lib:/usr/lib:/usr/local/lib`, but the path may vary between linux distros; the point is that you don’t depend on exact path, this is handled by libc) as any other system library – grammar libraries are expected to be named `libtree-sitter-<lang>.so` (diffsitter’s approach).

Or you can look up grammar libraries directly in some predefined directory (e.g. `/usr/lib/tree-sitter` or some app-specific dir) named as `<lang>.so` (helix’s approach).

The former is better, but both are acceptable for Linux distros – in the latter case, we can just symlink `<your-directory-for-grammars>` to e.g. `/usr/lib/tree-sitter` where system-wide grammars are installed by the system’s package manager (and used by all programs that use tree-sitter). These approaches are not mutually exclusive – you can support both.

Also, you can continue to provide some grammars with tree-grepper (as both diffsitter and helix do, they use git submodules) for the convenience of the users that build tree-grepper themselves and their distro doesn’t provide tree-sitter grammars yet. The aim is to _allow_ package maintainers to package grammars separately and give users the freedom to choose what grammars they will install.

**Comment by Wilfred** [comment]
Ooh, are you interested in packaging difftastic @jirutka? That's be really neat.

This seems like a reasonable feature to add, but it'll require a few fiddly details:

(1) Difftastic doesn't just use the parser's C library, it uses the syntax highlighting file(s) (usually named `queries/highlights.scm`) too.

(2) For each language, difftastic also has configuration that specifies when to use the parser (e.g. file extensions) and how to use it (which AST nodes map to diffsitter's AST).

(3) Difftastic doesn't provide any configuration files yet, just a few CLI options and environment variables. Typically difftastic is invoked from git or hg rather than by the user, so I've generally preferred environment variables so far. I'm not sure what a good configuration system looks like for difftastic, so it'll need some design work.

(4) In a few cases I've forked parsers. I try to upstream changes but some parsers are essentially abandoned.

**Comment by jirutka** [comment]
> Ooh, are you interested in packaging difftastic @jirutka? That'd be really neat.

Yes, I’m Alpine Linux developer and difftastic looks like a very promising and useful project.

> (1) Difftastic doesn't just use the parser's C library, it uses the syntax highlighting file(s) (usually named queries/highlights.scm) too.

So is e.g. Helix. However, Helix is not the best inspiration in this matter, because they have vendored all queries and it’s unclear which are just copied from the respective grammar projects and which ones are modified. They install the queries into `$HELIX_RUNTIME/queries/<lang>` where `$HELIX_RUNTIME` is typically `/usr/share/helix/runtime` when installed from a package provided by a (Linux) distribution.

On Alpine Linux, we package queries (`.scm` files) together with compiled tree-sitter grammars and install them into `/usr/share/tree-sitter/queries/<lang>/`.

The ideal solution would be to add support for searching queries in multiple base directories (per-site configurable or hard-coded):

1. system directory with stock queries provided with the grammars (`/usr/share/tree-sitter/queries` on Linux)
2. (system) directory with queries provided by difftastic (e.g. `/usr/share/difftastic/queries` on Linux)
3. user directory for queries provided/installed by user (`difftastic/queries` under [`$XDG_DATA_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) on Linux)

The first two would be basically intended for the distributions packaging difftastic. In the case of manual installations, i.e. when the user builds (or downloads a tarball with prebuilt binary and other files) and installs difftastic themself, difftastic can provide an archive of bundled queries (both stock and custom; basically what you have now) and instruct the user (or provide a script fort hat) to copy them all to (3).

Note: `/usr/share` prefix shouldn’t be hard-coded, but built-time configurable by some mechanism in cargo, but I don’t remember how cargo handles this.

> (3) Difftastic doesn't provide any configuration files yet, just a few CLI options and environment variables. Typically difftastic is invoked from git or hg rather than by the user, so I've generally preferred environment variables so far. I'm not sure what a good configuration system looks like for difftastic, so it'll need some design work.

Making the paths user-configurable would great, but I don’t think it’s necessary – convention over configuration would be IMO sufficient. The system locations, primarily used by distributions, can be just build-time configurable (even a constant in a rust source file with some sensible default which can be easily patched if needed would be sufficient, at least for Alpine).

For the user locations, there’s a convenient cross-platform Rust crate [dirs](https://crates.io/crates/dirs) that abstracts from the details of each platform.

> (4) In a few cases I've forked parsers. I try to upstream changes but some parsers are essentially abandoned.

Upstreaming is the way to go. :+1: If some parser is abandoned, you can promote your fork as an up-to-date replacement and package maintainers will catch on (unless there are several competing forks and it’s unclear which one should be preferred).

**Comment by Wilfred** [comment]
https://github.com/Wilfred/difftastic/issues/438#issuecomment-1368348008 discusses the different approaches being used in the tree-sitter ecosystem for this.

**Comment by Wilfred** [comment]
https://crates.io/crates/tree-sitter-loader is a recent project that could really help here.

**Comment by Wilfred** [comment]
A bunch of useful comments on duplicate issue #438.

**Comment by tingerrr** [comment]
I've recently written my own grammar for a DSL at my day job and adding it to a local clone of difftastic was fairly easy using a path dependency to the Rust bindings of it, but maintaining a fork like that is fairly annoying.

> So is e.g. Helix. However, Helix is not the best inspiration in this matter, because they have vendored all queries and it’s unclear which are just copied from the respective grammar projects and which ones are modified.

I've noticed a trend of many in-house solutions to the same problems in different projects with seemingly no coordination between them. As of yet, I'm not aware of any easy way to write a tree sitter grammar which:
- Contains blessed queries for multiple consumers. nvim-tree-sitter, helix, zed, and `tree-sitter highlight` all seem to be using slightly different queries which are incompatible, tree sitter doesn't make it easy to generate them so you end up maintaining them by hand or writing your own generation script.
- Doesn't track a bunch of useless parser files which could be generated by tree-sitter on demand. Zed requires the files to be tracked, because plugins don't run the generation step after checking out the grammar repo, so if try to they build the WASM module of the grammar it's missing the `parser.c` and relate files and can't be loaded.

As you noted `tree-sitte-loader` might be of interest, the helix maintainers have also apparently built a package manager for tree sitter grammars, but so far I'm not sure how it works or if they even encourage users to use it. The repo for that is here: https://github.com/helix-editor/tree-house, though it seems that they changed the description to no longer claim that `skidder` is a package manager.

With all that in mind, perhaps difftastic should attempt to coordinate with the other projects to avoid having users consume tree sitter grammars in yet anoher way.

**Comment by abcdw** [comment]
Yeah, in GNU Guix we also have difftastic packaged with vendored grammars, so we can neither utilize other guix-packaged tree-sitter grammars, nor user-provided ones. Would be really great to have an ability to load them in runtime. BTW, thank you for a great tool, I'm following your project for a long time (even before tree-sitter I guess), but had a chance to actually enjoy it only recently:

- https://youtu.be/pARy5XnLHKQ
- https://fosstodon.org/@abcdw/116160058857666288

Thank you for your hard work, it is inspiring and exceptionally useful!  I think that structural diffs is a future. Also, I plan to work on a Scheme grammar to better support Guix and Guile source code and hope to use difftastic daily :)
